### PR TITLE
PRO-168: dev command

### PR DESCRIPTION
**Pull request by BetaHuhn**

### CODE REVIEW

File: cmd/dev.go [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+	go checkVersion(c)
+
+	var err error
+
+	// grab to port for the proxy server from the env or default to 8080
+	serverPort := os.Getenv("PORT")
+	if serverPort == "" {
+		serverPort = "8080"
+	}
+	hostname := fmt.Sprintf("http://localhost:%s", serverPort)
```
Comment (CyberL1) [inline cmd/dev.go:185]: the `DETA_SPACE_APP_HOSTNAME` env variable in micros doesn't inclue the protocol prefix
```suggestion
	hostname := fmt.Sprintf("localhost:%s", serverPort)
```

**Comment by pomdtr** [comment]
What do you thing of generating a docker-compose config ? It would allow to use nginx as the reverse proxy, and to be able to monitor the logs of all of the launched micros.

**Comment by BetaHuhn** [comment]
> What do you thing of generating a docker-compose config, and having space dev be an alias for docker-compose up ? It would allow to use nginx as the reverse proxy, and to be able to monitor the logs of all of the launched micros. The env could also be passed from an .env file, and there would not be any port collision possible.

This would require devs to have docker/docker-compose installed though right? Ideally no additional setup is required apart from installing the Space CLI to be able to use `space dev`. We want the experience to be as simple as possible.

What you are talking about might make more sense as a community maintained project/add-on.

**Comment by pomdtr** [comment]
I feel like this will hardly be true in practice. For example running python micros will require to have unicorn or another asgi runner installed.

**Comment by BetaHuhn** [comment]
> I feel like this will hardly be true in practice. For example running python micros will require to have unicorn or another asgi runner installed, and some micros might event install some additional requirements in the command array.

But you will likely already have them installed if you are building an app with a Python micro. The command is more intended for devs to be able to easily develop their own Space app by automatically generating the right keys so they can use Base and Drive and partially emulating the Space environment to some extent.

Allowing others to run the app locally to be able to contribute is a separate issue which could be solved by a combination of using Docker and the Space CLI. The CLI could handle the keys while docker handles running the application for example.

**Comment by pomdtr** [comment]
I feel like these two issues are linked, but I get that docker might introduce some frictions for developpers.

Btw is there a way to generate a project key from the UI ?

**Comment by BetaHuhn** [comment]
> Btw is there a way to generate a project key from the deta.space website ?

Project keys no longer exist, you need to use data keys instead. To generate a data key for a project go to the "Develop" page and then the "Data" tab. We are aware that keys are confusing at the moment and are working on simplifying it as well as updating the documentation of the SDKs.

**Comment by pomdtr** [comment]
Make senses, I was actually mostly confused by the deta.space doc: https://deta.space/docs/en/basics/data#developing-with-base-and-drive

**Comment by abdelhai** [comment]
i feel we could provide both options. a lightweight `space dev` and a docker-backed `space dev full`...

**Comment by pomdtr** [comment]
I agree that their should be an options to run the project locally. However I don't think that grouping all of the micros in a single command is the right way to go.

What do you think of autogenerating tasks files for common ides (ex: vscode, intellij, neovim) ? It would allow to have the output of each micros in a separate terminal.

**Comment by BetaHuhn** [comment]
> What do you think of autogenerating tasks files for common ides (ex: vscode tasks) ? It would allow to have the output of each micros in a separate terminal, and a good dev experience.

Again this could be useful, but I don't think it makes sense for this to be the default `space dev` experience. The command should be fully independent of the dev's IDE and other dependencies.

> However I don't think that grouping all of the micros in a single command is the right way to go.

Why not? Is the main concern combining the logs?

**Comment by pomdtr** [comment]
Yes, at least you should have a way to discriminate beetween the logs of multiple micros.

**Comment by BetaHuhn** [comment]
> I feel like the user should be able to launch each micros separately, or even just launch the reverse proxy.

I like the idea of launching Micros separately as the pieces for it are already there and I could see it being useful when integrating with other workflows.

What would be a use case of running the reverse proxy separately?

**Comment by pomdtr** [comment]
Basically it would allow the user reproduce the space dev command, but better integrate it with it's workflow.

If I take the vscode tasks example:

```
{
    "version": "2.0.0",
    "tasks": [
        {
            "label": "micro 1",
            "command": "space micro start <micro-1>"
        },
        {
            "label": "micro 2",
            "command": "space micro start <micro-2>"
        },
        {
            "label": "Space Dev",
            "command": "space rproxy",
            "dependsOn": [
                "micro 1",
                "micro 2"
            ]
        }
    ]
}
```

**Comment by pomdtr** [comment]
I have experimented a bit with deta-dev, and I'm quite close to a working solution (see: https://github.com/pomdtr/space-dev).

However my proxy is too simplistic, so it does not work for application that reference their own assets (ex: next apps).

**Pull request closed without merging**